### PR TITLE
Create layout interactive elements

### PR DESCRIPTION
**Pull request by brean**
New, small Example for Dynamic Buttons, see https://github.com/Microsoft/MixedRealityToolkit-Unity/issues/1137

### CODE REVIEW

File: Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        /// modify the TargetGroup data
+        /// </summary>
+        public void NoteTypeButtonSelected()
+        {
+            if (SourceSet.SelectedIndices.Count == 0)
+            {
+                return;
+            }
+            int interactivePos = SourceSet.SelectedIndices[0];
+            string title = SourceSet.Interactives[interactivePos].gameObject.GetComponent<LabelTheme>().Default;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs]: Anytime you use `GetComponent<T>` you should always check if it's null before using it's returned value.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                // layouting
+                int j = i % rows;
+                Collider collider = interactive.gameObject.GetComponent<Collider>();
+                Vector2 Distance = new Vector2(
+                    collider.bounds.size.x + Offsets.x,
+                    collider.bounds.size.y + Offsets.y
+                );
+                interactive.gameObject.transform.localPosition = new Vector3(
+                    ((i / rows) - ((columns - 1) / 2f)) * Distance.x,
+                    -(j - (rows - 1) / 2f) * Distance.y);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: Instad of `value / 2` we can use `value * 0.5f`

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveSet.cs [hunk 3]
```diff
@@ -22,7 +22,7 @@ public class InteractiveSet : MonoBehaviour
         public SelectionType Type = SelectionType.single;
 
         [Tooltip("Interactives that will be managed by this controller")]
-        public InteractiveToggle[] Interactives;
+        public List<InteractiveToggle> Interactives;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveSet.cs]: What's the advantage of changing this to a list?

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        /// shortcut to set title
+        /// (assuming this Interactive has a LabelTheme attached to it)
+        /// </summary>
+        /// <param name="title"></param>
+        public void SetTitle(string title)
+        {
+            LabelTheme lblTheme = gameObject.GetComponent<LabelTheme>();
+            if (lblTheme == null)
+            {
+                Debug.LogError("No LabelTheme attached to this Interactive");
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: We could probably add a component requirement, and always get the component if null, and remove the debug.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs (hunk 2 above)
Comment (brean) [reply Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: I think value / 2 is better for (human) readability and I don't think the CPU will mind (its not like we are doing this loop for thousands of objects). Or is there a coding convention I am missing?

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs (hunk 4 above)
Comment (brean) [reply Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: agree.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: Idk, as a human, I've always hated fractions.

as for CPU perf, gotta remember we're also targeting HoloLens, and it's a mobile device and every little bit counts.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveSet.cs (hunk 3 above)
Comment (brean) [reply Assets/HoloToolkit-Examples/UX/Scripts/InteractiveSet.cs]: It's easier to remove items in a list, see the "RemoveInteractive"-method in the same file. If the data changes I don't delete all Interactives and their prefabs, only the delta and update all items afterwards (recycling objects).

**Comment by brean** [comment]
@StephenHodgson I implemented your feedback. Please review.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs [hunk 5]
```diff
@@ -102,6 +99,10 @@ public void UpdateData()
                 // layouting
                 int j = i % rows;
                 Collider collider = interactive.gameObject.GetComponent<Collider>();
+                if (collider == null)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: Should we also add `Collider` to the list of required components?

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs [hunk 6]
```diff
@@ -58,7 +59,10 @@ void CreateInteractives()
             {
                 GameObject PrefabInst = Instantiate(InteractivePrefab, gameObject.transform) as GameObject;
                 InteractiveToggle InterInst = PrefabInst.GetComponent<InteractiveToggle>();
-                Interactives.Add(InterInst);
+                if (InterInst != null)
+                {
+                    Interactives.Add(InterInst);
+                }
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: Should we throw a warning if our prefab doesn't have the `InteractiveToggle` telling the dev to add one to their prefab?

File: Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs [hunk 7]
```diff
@@ -46,9 +46,12 @@ public void NoteTypeButtonSelected()
...
             int interactivePos = SourceSet.SelectedIndices[0];
-            string title = SourceSet.Interactives[interactivePos].gameObject.GetComponent<LabelTheme>().Default;
-            TargetGroup.Titles = Data[title];
-            TargetGroup.UpdateData();
+            LabelTheme label = SourceSet.Interactives[interactivePos].gameObject.GetComponent<LabelTheme>();
+            if (label != null)
+            {
+                TargetGroup.Titles = Data[label.Default];
+                TargetGroup.UpdateData();
+            }
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs]: Should we add a warning if our object doesn't have a `LabelTheme` telling our dev to add one?

Comment (brean) [reply Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs]: Yes. The Dev would most likely implement his own "DataProvider", where he gets some data from an external source and than acts on it accordingly, putting that data into InteractiveSets/InteractiveGroups as he likes. So I also added some documentation that descibes it a bit more.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs (hunk 5 above)
Comment (brean) [reply Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: That would be the Collider in InteractiveToggle. I can attach the InteractiveToggle-script to any GameObject and have the collider in some child of that GameObject, so I don't think having it as required component is a good idea... we can however ignore the collider boundary size for layouting if it is not present. If you don't have a collider attached directly the user can define the distance between the InteractiveToggles by setting the Offsets-values. Alternatively I could iterate over transform.children recursively and calculate the biggest bounding box, but that seems a bit overkill.

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs (hunk 6 above)
Comment (brean) [reply Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: Yes.

**Comment by brean** [comment]
@StephenHodgson I am done integrating your feedback, please review this.

File: Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs [hunk 8]
```diff
@@ -4,14 +4,22 @@
...
         public InteractiveGroup TargetGroup;
 
         public InteractiveSet SourceSet;
 
-        // some test data - imagine this comming from a web-service or some input menu
-        public Dictionary<string, List<string>> Data = new Dictionary<string, List<string>>
+        // some test data - imagine this comming from a web-service 
+        // or some input menu
+        public Dictionary<string, List<string>> Data = 
+            new Dictionary<string, List<string>>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs]: Let's keep all of this on one line, and updated the comments with summary tags

File: Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    public class NoteDataProvider : MonoBehaviour
+    {
+        public InteractiveGroup TargetGroup;
+
+        public InteractiveSet SourceSet;
+
+        // some test data - imagine this comming from a web-service 
+        // or some input menu
+        public Dictionary<string, List<string>> Data = 
+            new Dictionary<string, List<string>>
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/ExampleData/NoteDataProvider.cs]: Let's keep all of this on one line, and updated the comments with summary tags

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+        [Tooltip("Data to fill the InteractiveSet.")]
+        public List<string> Titles = new List<string>();
+
+        [Tooltip("number of elements after which we start a new column")]
+        public int Rows = 3;
+
+        public Vector2 Offsets = new Vector2(0.00f, 0.00f);
+
+        void Start()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: nit: private

File: Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        {
+            RemoveInteractives();
+        }
+ 
+
+        /// <summary>
+        /// remove unused Interactives from scene
+        /// </summary>
+        /// <param name="keep">number of Iteractives that will NOT be deleted</param>
+        void RemoveInteractives(int keep = 0)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit-Examples/UX/Scripts/InteractiveGroup.cs]: nit: private

**Comment by brean** [comment]
@StephenHodgson done.

**Comment by brean** [comment]
I like to see this merged and it has been approved by @StephenHodgson  and @DDReaper . Is there anything I can do to speed up the process? Should I take a deeper look at the Error messages @DDReaper mentioned?

**Comment by DDReaper** [comment]
The issue is a separate one, not related to this PR, so it doesn't need fixing here.

**Pull request merged**